driver-core/src/main/com/mongodb/internal/connection/AwsAuthenticator.java

```diff
@@ -223,20 +223,18 @@ private String getPassword() {
 
         @Nullable
         private String getSessionToken() {
-            String token = credential.getMechanismProperty("AWS_SESSION_TOKEN", null);
-            if (credential.getPassword() == null || credential.getUserName() == null) {
-                if (token == null) {
-                    if ((token = System.getenv("AWS_SESSION_TOKEN")) == null) {
-                        token = BsonDocument
-                                .parse(getHttpResponse())
-                                .getString("Token")
-                                .getValue();
-                    }
-                } else {
-                    throw new IllegalArgumentException("The connection string contains auth properties and no username and password");
-                }
+            if (credential.getPassword() != null && credential.getUserName() != null) {
+                return credential.getMechanismProperty("AWS_SESSION_TOKEN", null);
             }
-            return token;
+
+            if (System.getenv("AWS_SECRET_ACCESS_KEY") != null && System.getenv("AWS_ACCESS_KEY_ID") != null) {
+                return System.getenv("AWS_SESSION_TOKEN");
```

Comment:
OK, so if no credentials in the URI, we ignore the mechanism property as well and prefer the environment variable.  That seems strange to me, but it does seem correct according to spec.  Are you pretty confident this is what users would expect?

Reply:
I'm open to either behavior but I think it can get really confusing if we fall though. For example, if a user tries to authenticate via the URI and for some reason fails to add a password. Then we'd fall through to the env vars, which they did not set. Then attempt to check if we're on an ECS container, which we won't be. Then we'd try to check the EC2 endpoint and fail with an IOException. The user would then wonder why our code is attempting to contact an EC2 endpoint when they've explicitly (tried) to give us credentials.